bigtable-client-core-parent/bigtable-hbase/src/main/java/com/google/cloud/bigtable/hbase/wrappers/veneer/DataClientVeneerApi.java

```diff
@@ -228,6 +241,147 @@ protected void onCompleteImpl() {
+    private Boolean hasMore = true;
+    private final Queue<Result> buffer;
+    private final Query.QueryPaginator paginator;
+    private final int refillSegmentWaterMark;
+
+    private final BigtableDataClient dataClient;
+
+    private final long maxSegmentByteSize;
+
+    private long currentByteSize = 0;
+
+    private @Nullable Future<List<Result>> future;
+    private GrpcCallContext scanCallContext;
+
+    PaginatedRowResultScanner(
+        Query.QueryPaginator paginator,
+        BigtableDataClient dataClient,
+        long maxSegmentByteSize,
+        GrpcCallContext scanCallContext) {
+      this.maxSegmentByteSize = maxSegmentByteSize;
+
+      this.paginator = paginator;
+      this.dataClient = dataClient;
+      this.buffer = new ArrayDeque<>();
+      this.refillSegmentWaterMark =
+          (int) Math.max(1, paginator.getPageSize() * WATERMARK_PERCENTAGE);
+      this.scanCallContext = scanCallContext;
+      this.future = fetchNextSegment();
+    }
+
+    @Override
+    public Result next() {
+      try (Context ignored = scannerResultTimer.time()) {
+        if (this.future != null && this.future.isDone()) {
+          this.consumeReadRowsFuture();
+        }
+        if (this.buffer.size() < this.refillSegmentWaterMark && this.future == null && hasMore) {
+          future = fetchNextSegment();
+        }
+        if (this.buffer.isEmpty() && this.future != null) {
+          this.consumeReadRowsFuture();
+        }
+        Result result = this.buffer.poll();
+        if (result != null) {
+          scannerResultMeter.mark();
+          currentByteSize -= Result.getTotalSizeOfCells(result);
+        }
+        return result;
+      }
+    }
+
+    @Override
+    public void close() {
+      if (this.future != null) {
+        this.future.cancel(true);
+      }
+    }
+
+    public boolean renewLease() {
+      return true;
+    }
+
+    private Future<List<Result>> fetchNextSegment() {
+      SettableFuture<List<Result>> resultsFuture = SettableFuture.create();
+
+      dataClient
+          .readRowsCallable(RESULT_ADAPTER)
+          .call(
+              paginator.getNextQuery(),
+              new ResponseObserver<Result>() {
+                private StreamController controller;
+                List<Result> results = new ArrayList();
+
+                @Override
+                public void onStart(StreamController controller) {
+                  this.controller = controller;
+                }
+
+                @Override
+                public void onResponse(Result result) {
+                  // calculate size of the response
+                  currentByteSize += Result.getTotalSizeOfCells(result);
+                  results.add(result);
+                  if (result != null && result.rawCells() != null) {
+                    lastSeenRowKey = RESULT_ADAPTER.getKey(result);
+                  }
+
+                  if (currentByteSize > maxSegmentByteSize) {
+                    controller.cancel();
```

Comment:
let's backport the bug we caught in beam https://github.com/apache/beam/commit/a221f98a5f46be985afcb98a65fcec3b46b81f92